com.unity.render-pipelines.core/Runtime/Volume/KeyframeUtility.cs

```diff
@@ -0,0 +1,294 @@
+            {
+                // In this case the obvous solution would be to return the rhsCurve. BUT (!) the lhsCurve and rhsCurve are different. This function is
+                // called by:
+                //      stateParam.Interp(stateParam, toParam, interpFactor);
+                //
+                // stateParam (lhsCurve) is a temporary in/out parameter, but toParam (rhsCurve) might point to the original component, so it's unsafe to
+                // change that data. Thus, we need to copy the keys from the rhsCurve to the lhsCurve instead of returning rhsCurve.
+                KeyframeUtility.ResetAnimationCurve(ref lhsAndResultCurve);
+
+                for (int i = 0; i < rhsCurve.length; i++)
+                {
+                    lhsAndResultCurve.AddKey(rhsCurve.keys[i]);
+                }
+                lhsAndResultCurve.postWrapMode = rhsCurve.postWrapMode;
+                lhsAndResultCurve.preWrapMode = rhsCurve.preWrapMode;
+            }
+            else
+            {
+                // Note: If we reached this code, we are guaranteed that both lhsCruve and rhsCurve are valid with at least 1 key
+
+                // first, figure out the start and end time to include both curves
+                var lhsCurveKeys = lhsAndResultCurve.keys;
+                var rhsCurveKeys = rhsCurve.keys;
+
+                float startTime = Mathf.Min(lhsCurveKeys[0].time, rhsCurveKeys[0].time);
+                float endTime = Mathf.Max(lhsCurveKeys[lhsAndResultCurve.length - 1].time, rhsCurveKeys[rhsCurve.length - 1].time);
+
+                // we don't know how many keys the resulting curve will have (because we will compact keys that are at the exact
+                // same time), but in most cases we will need the worst case number of keys. So allocate the worst case.
+                int maxNumKeys = lhsAndResultCurve.length + rhsCurve.length;
+                int currNumKeys = 0;
+                NativeArray<Keyframe> dstKeys = new NativeArray<Keyframe>(maxNumKeys, Allocator.Temp);
+
+                int lhsKeyCurr = 0;
+                int rhsKeyCurr = 0;
+
+                while (lhsKeyCurr < lhsCurveKeys.Length || rhsKeyCurr < rhsCurveKeys.Length)
+                {
+                    // the index is considered invalid once it goes off the end of the array
+                    bool lhsValid = lhsKeyCurr < lhsCurveKeys.Length;
+                    bool rhsValid = rhsKeyCurr < rhsCurveKeys.Length;
+
+                    // it's actually impossible for lhsKey/rhsKey to be uninitialized, but have to
+                    // add initialize here to prevent compiler erros
+                    var lhsKey = new Keyframe();
+                    var rhsKey = new Keyframe();
+                    if (lhsValid && rhsValid)
+                    {
+                        lhsKey = GetKeyframeAndClampEdge(lhsCurveKeys, lhsKeyCurr);
+                        rhsKey = GetKeyframeAndClampEdge(rhsCurveKeys, rhsKeyCurr);
```

Comment:
If you apply my suggestion on the `GetKeyframeAndClampEdge` you will be able to initialize directly `lhsKey` and `rhsKey` by calling the method, and reduce the cyclomatic complexity of the loop.

Reply:
How exactly would that work? GetKeyframeAndClampEdge() needs to adjust the start and end time based on the segment start, whereas the other calls to GetKeyframeAndClampEdge() need to get the key within the current curve (not including segment start and end).

Reply:
I didn't realize that they were two different usages for it :)